src/core/qgsrendercontext.cpp

```diff
@@ -736,7 +745,22 @@ QList<QPainterPath> QgsRenderContext::symbolLayerClipPaths( const QString &symbo
 
 void QgsRenderContext::addSymbolLayerClipGeometry( const QString &symbolLayerId, const QgsGeometry &geometry )
 {
-  mSymbolLayerClippingGeometries[ symbolLayerId ].append( geometry );
+  if ( geometry.isMultipart() )
+  {
+    mSymbolLayerClippingGeometries[ symbolLayerId ].append( geometry.asGeometryCollection() );
+  }
+  else
+  {
+    mSymbolLayerClippingGeometries[ symbolLayerId ].append( geometry );
+  }
+}
+
+bool QgsRenderContext::symbolLayerHasClipGeometries( const QString &symbolLayerId ) const
```

Comment:
Is it only for test ? Could we not use `symbolLayerClipGeometries` ?

Reply:
This is used in a follow up optimisation which I'll submit shortly as a separate PR.